salt/beacons/adb.py

```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+'''
+Beacon to emit adb device state changes for Android devices
+'''
+
+# Import Python libs
+from __future__ import absolute_import
+import logging
+
+# Salt libs
+import salt.utils
+
+log = logging.getLogger(__name__)
+
+__virtualname__ = 'adb'
+
+last_state = {}
+last_state_no_devices = False
+
+
+def __virtual__():
+    which_result = salt.utils.which('adb')
+    if which_result is None:
+        return False
+    else:
+        return __virtualname__
+
+
+def validate(config):
+    '''
+    Validate the beacon configuration
+    '''
+    # Configuration for adb beacon should be a dictionary with states array
+    if not isinstance(config, dict):
+        log.info('Configuration for adb beacon must be a dict.')
+        return False
+    elif "states" not in config.keys():
+        log.info('Configuration for adb beacon must include a states array.')
+        return False
+    else:
+        states = ['offline', 'bootloader', 'device', 'host', 'recovery', 'no permissions',
+                  'sideload', 'unauthorized', 'unknown', 'missing']
+        if any(s not in states for s in config["states"]):
+            log.info('Need a one of the following adb states: {0}'.format(', '.join(states)))
+            return False
+    return True
+
+
+def beacon(config):
+    '''
+    Emit the status of all devices returned by adb
+
+    Specify the device states that should emit an event,
+    there will be an event for each device with the
+    event type and device specified.
+
+    .. code-block:: yaml
+
+        beacons:
+          adb:
+            - states:
+                - offline
+                - unauthorized
+                - missing
+            - no_devices_event: True
+            - battery_low: 25
+
+    '''
+    global last_state_no_devices
```

Comment:
We typically discourage the use of globals. Is there a way to refactor this code so that this is not necessary?